lib/membrane_h264_plugin/parser/au_splitter.ex

```diff
@@ -151,76 +151,59 @@ defmodule Membrane.H264.Parser.AUSplitter do
-
-        # in case of a nalu holding the keyframe, return true
-        # this one wasn't specified in the documentation, however it seems logical to do so
-        # and not doing that has made the application crash
-
-        nalu.type == :idr ->
-          true
-
-        nalu.parsed_fields.frame_num != last_nalu.parsed_fields.frame_num ->
-          true
-
-        nalu.parsed_fields.pic_parameter_set_id != last_nalu.parsed_fields.pic_parameter_set_id ->
-          true
-
-        Map.has_key?(nalu.parsed_fields, :field_pic_flag) and
-          Map.has_key?(last_nalu.parsed_fields, :field_pic_flag) and
-            nalu.parsed_fields.field_pic_flag != last_nalu.parsed_fields.field_pic_flag ->
-          true
-
-        Map.has_key?(nalu.parsed_fields, :field_pic_flag) and
-          Map.has_key?(last_nalu.parsed_fields, :field_pic_flag) and
-          nalu.parsed_fields.field_pic_flag == 1 and
-          last_nalu.parsed_fields.field_pic_flag == 1 and
-            nalu.parsed_fields.bottom_field_flag != last_nalu.parsed_fields.bottom_field_flag ->
-          true
-
-        nalu.parsed_fields.nal_ref_idc != last_nalu.parsed_fields.nal_ref_idc and
-            Enum.any?(
-              [nalu.parsed_fields.nal_ref_idc, last_nalu.parsed_fields.nal_ref_idc],
-              &(&1 == 0)
-            ) ->
-          true
-
-        nalu.parsed_fields.pic_order_cnt_type == 0 and
-          last_nalu.parsed_fields.pic_order_cnt_type == 0 and
-            (nalu.parsed_fields.pic_order_cnt_lsb != last_nalu.parsed_fields.pic_order_cnt_lsb or
-               Map.get(nalu.parsed_fields, :delta_pic_order_cnt_bottom) !=
-                 Map.get(last_nalu.parsed_fields, :delta_pic_order_cnt_bottom)) ->
-          true
-
-        nalu.parsed_fields.pic_order_cnt_type == 1 and
-          last_nalu.parsed_fields.pic_order_cnt_type == 1 and
-            (Bunch.Access.get_in(nalu.parsed_fields, [:delta_pic_order_cnt, 0]) !=
-               Bunch.Access.get_in(last_nalu.parsed_fields, [:delta_pic_order_cnt, 0]) or
-               Bunch.Access.get_in(nalu.parsed_fields, [:delta_pic_order_cnt, 1]) !=
-                 Bunch.Access.get_in(last_nalu.parsed_fields, [:delta_pic_order_cnt, 1])) ->
-          true
-
-        # The following condition to be checked is also specified in the documentation: "IdrPicFlag is equal to 1 for
-        # both and idr_pic_id differs in value".
-        # At the same time, it seems that it describes the situation, that we are having two different IDR frames - and
-        # that should imply, that their frame_num should differ, which is already checked in the second condition
-
-        true ->
-          false
-      end
-    else
-      false
-    end
+  defguardp frame_num_differs(a, b) when a.frame_num != b.frame_num
+
+  defguardp pic_parameter_set_id_differs(a, b)
+            when a.pic_parameter_set_id != b.pic_parameter_set_id
+
+  defguardp field_pic_flag_differs(a, b) when a.field_pic_flag != b.field_pic_flag
```

Comment:
That one might be problematic. The specification states that: "When field_pic_flag is not present it shall be inferred to be equal to 0."
Suppose that `a` doesn't have `field_pid_flag` and `b` has a `field_pid_flag=1`. In that situation, the guard should return `true`, as the `field_pid_flag` should be inferred to 0, and `0 != 1`. With current implementation, the guard will return `false`

Reply:
It can be solved by adding `field_pid_flag: 0` to the default fields list here: https://github.com/membraneframework/membrane_h264_plugin/blob/f94cd90c7810e4e5169cb6d838df1b513e1891be/lib/membrane_h264_plugin/parser/nalu_parser/schemes/slice.ex#L6

Reply:
The same might be necessary for other fields, which might not be present, but for which the specification defines the default values. As far as I can see all the default values for fields in `is_new_primary_coded_vcl_nalu` are simply inferred if some fields are not present in the stream, but it requires some attention, as in the specification there are cases where the inferred value depends on the "context", e.g.:

"the picture shall be inferred to have had frame_num equal to 0 for all subsequent use in the decoding process, except as specified in subclause 7.4.1.2.4. " 
 
It suggests, that independent of `frame_num` value read from stream, it should be set to `0`, except 7.4.1.2.4 (on which we base our `is_new_primary_coded_vcl_nalu` implementation), where we should use the original value. Fortunately, the `frame_num` is always present in the stream, so no actions are required, but I am just raising awareness that weird stuff might happen :D